ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/handler/AbortedTxnCleaner.java

```diff
@@ -0,0 +1,168 @@
+    long abortedTimeThreshold = HiveConf
+              .getTimeVar(conf, HiveConf.ConfVars.HIVE_COMPACTOR_ABORTEDTXN_TIME_THRESHOLD,
+                      TimeUnit.MILLISECONDS);
+    List<AcidTxnInfo> readyToCleanAborts = txnHandler.findReadyToCleanForAborts(abortedTimeThreshold, abortedThreshold);
+
+    if (!readyToCleanAborts.isEmpty()) {
+      return readyToCleanAborts.stream().map(ci -> ThrowingRunnable.unchecked(() ->
+                      clean(ci, ci.txnId > 0 ? ci.txnId : Long.MAX_VALUE, metricsEnabled)))
+              .collect(Collectors.toList());
+    }
+    return Collections.emptyList();
+  }
+
+  private void clean(AcidTxnInfo info, long minOpenTxn, boolean metricsEnabled) throws MetaException {
+    LOG.info("Starting cleaning for {}", info);
+    PerfLogger perfLogger = PerfLogger.getPerfLogger(false);
+    String cleanerMetric = MetricsConstants.COMPACTION_CLEANER_CYCLE + "_";
+    try {
+      if (metricsEnabled) {
+        perfLogger.perfLogBegin(AbortedTxnCleaner.class.getName(), cleanerMetric);
+      }
+      Table t;
+      Partition p = null;
+      t = metadataCache.computeIfAbsent(info.getFullTableName(), () -> resolveTable(info.dbname, info.tableName));
+      if (isNull(t)) {
+        // The table was dropped before we got around to cleaning it.
+        LOG.info("Unable to find table {}, assuming it was dropped.", info.getFullTableName());
+        txnHandler.markCleanedForAborts(info);
+        return;
+      }
+      if (MetaStoreUtils.isNoCleanUpSet(t.getParameters())) {
+        // The table was marked no clean up true.
+        LOG.info("Skipping table {} clean up, as NO_CLEANUP set to true", info.getFullTableName());
+        return;
+      }
+      if (!isNull(info.partName)) {
+        p = resolvePartition(info.dbname, info.tableName, info.partName);
+        if (isNull(p)) {
+          // The partition was dropped before we got around to cleaning it.
+          LOG.info("Unable to find partition {}, assuming it was dropped.",
+                  info.getFullPartitionName());
+          txnHandler.markCleanedForAborts(info);
+          return;
+        }
+        if (MetaStoreUtils.isNoCleanUpSet(p.getParameters())) {
+          // The partition was marked no clean up true.
+          LOG.info("Skipping partition {} clean up, as NO_CLEANUP set to true", info.getFullPartitionName());
+          return;
+        }
+      }
+
+      String location = CompactorUtil.resolveStorageDescriptor(t,p).getLocation();
+      info.runAs = TxnUtils.findUserToRunAs(location, t, conf);
+      abortCleanUsingAcidDir(info, location, minOpenTxn);
+
+    } catch (Exception e) {
+      LOG.error("Caught exception when cleaning, unable to complete cleaning of {} due to {}", info,
+              e.getMessage());
+      throw new MetaException(e.getMessage());
+    } finally {
+      if (metricsEnabled) {
+        perfLogger.perfLogEnd(AbortedTxnCleaner.class.getName(), cleanerMetric);
```

Comment:
I think it does not worth doing it

1. It's a metrics, not a critical component or code part
2. adding a retry (especially with back-off would possibly corrupt the metrics value, as the perfLogEnd call would not be in sync with the complete time of the operation it measures

Reply:
@veghlaci05 My bad, when I mentioned retry I commented on the finally block, what I meant to do is trigger a conversation around retry mechanism for the abortCleanUsingAcidDir not the metrics

Reply:
Until and unless metadata of the aborted transactions is not explicitly removed by AbortedTxnCleaner, the AbortedTxnCleaner always fetches aborted entries in TXN_COMPONENTS table. So there is an implicit functionality of a retry operation already in case of failures during abort cleanup.

Reply:
Thanks @SourabhBadhya, this works.

Reply:
Initiator had backoff based on retry count and time. Right now do we have any back-off? If no, can we reuse Initiator logic here?

Reply:
oh, we do, it's time based: handleCleanerAttemptFailure

Reply:
@deniskuzZ Yes there is no retry backoff as of now. Even if we want to implement this needs to be stored somewhere. Since we are not using COMPACTION_QUEUE, the only place I see this going is TXN_COMPONENTS table which I am not sure is desirable or not.

Reply:
can we place selected entries in compaction_queue in "r" state? if there would be some issue on s3, HMS might just die